### PR TITLE
Handle converting between interface implementation kinds

### LINKED ISSUE #2616: SyntaxGenerator.AsPublicInterfaceImplementation returns incorrect result for explicit implementations
Invoke `SyntaxGenerator.AsPublicInterfaceImplementation`(http://source.roslyn.io/Microsoft.CodeAnalysis.CSharp.Workspaces/R/f860f54a842df309.html) method on a method declaration which explicitly implements an interface method. For example, invoke it on the method implementation below:

```
public interface IGeneral
{
    object DoSomething();
}

public class ImplementsGeneral  : IGeneral
{
    object IGeneral.DoSomething() { return null; }
}
```

**Expected result:**
`public object DoSomething() { return null; }`

**Actual result:**
`public object IGeneral.DoSomething() { return null; }`

The resulting method has semantic error as modifiers are not allowed on explicitly implemented methods.

### PR DESCRIPTION
**Pull request by mattwar**
Fixes #2616 

Added new overloads of AsPublicInterfaceImplentation and AsPrivateInterfaceImplemenation that adds an extra last parameter that is the interface member's name.  If not supplied (null or calling via old overload) the method attempts to infer the member name from the existing member name (for backward compatibility) but this may not always be possible, so it is best to always supply the interface member name if you are modifying an declaration from existing code.

@mavasani @pilchie

### CODE REVIEW

**Comment by mattwar** [comment]
@mavasani @Pilchie  please review

File: src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb [hunk 1]
```diff
@@ -840,78 +840,89 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
...
-                Return method.WithSubOrFunctionStatement(
-                    method.SubOrFunctionStatement.WithImplementsClause(
-                        SyntaxFactory.ImplementsClause(SyntaxFactory.QualifiedName(type, interfaceMemberName))))
+        Private Function GetMemberName(declaration As SyntaxNode) As String
+            If GetImplementsClause(declaration) Is Nothing Then
+                Return GetName(declaration)
+            Else
+                Dim fullName = GetName(declaration)
+                If fullName IsNot Nothing Then
+                    Dim last = fullName.Split({"_"c}, StringSplitOptions.RemoveEmptyEntries).LastOrDefault()
```
Comment (mavasani) [inline src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb]: Do we need this heuristic? Given that `GetImplementsClause(declaration)` is non-null, can't we fetch the implemented member name from the implements clause?

**Comment by mavasani** [comment]
:+1: Thanks for all the prompt fixes!

File: src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb (hunk 1 above)
Comment (mattwar) [inline src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb]: Yes, that would probably be better.

**Pull request merged**